### DISCUSSION TITLE
Cant get file menu in AU export mac

### DISCUSSION THREAD

**Discussion by puffer3** [RomanKubiak/ctrlr#459] [Q&A] [answered]
Hello,

I have been working on more panels! Since my friends studio is so far away I thought I would develop a panel for one of my own synths so I could get more hands on experience using ctrlr in my workflow.  So here I have made a panel for my OBSx oberheim synthesizer newly equipped with the MagicBox midi/computer interface.  The panel is working well and sounds great.  Its amazing to be able to program this synth after only being able to use presets for so long! 

Unfortunately I am having some export issues I was hoping someone could clear up.  I cant seem to export an AU component that has a File Menu bar no matter how hard I try.  I am using 5.6.23 on mac BigSur 11.5.2 exporting an AU from ableton.  The panel and menu bar work great in the Ctrlr AU wrapper, but the problem here is that all my modulators are undefined so no automation is possible.

I tried exporting from another Mac laptop using both 5.6.23 and 5.3.198 and then putting the components in Ableton on my main computer, but the menu bar isnt exporting on that computer either! 

Of course the box for "is menu bar visible for this panel" is checked and all the hide menu bar options are unchecked.  Has anyone run into this issue?  Its frustrating bc I really want to try automating some parameters in ableton.

[OBSxMenuBarIssues.bpanelz.zip](https://github.com/RomanKubiak/ctrlr/files/8565946/OBSxMenuBarIssues.bpanelz.zip)

**Comment by damiensellier** [comment]
Hi, I never experienced this issue yet.
I can't see any issue on your panel that would hide the menu bar.
But I'm not sure about your exact problem, you are talking about "file menu bar".
When in restricted instance like AU or VST, I don't recall seeing this "File" section in the menu bar. There's just the "View", "Midi", and "Help" If I remember. 
It would be pretty useless anyway since it's meant for saving and exporting the current panel while in "Wrapper" mode.

So can you please clarify if the whole top menu bar is missing or just its "File" section?

If the whole bar is missing you can try to force the visibilty of the menu bar on startup with a global script for "called when the panel has finished loading":

panel:getPanelEditor():setProperty("uiPanelMenuBarVisible","1",true)
panel:getPanelEditor():setProperty("uiPanelMenuBarHideOnExport","0",true)

_but the problem here is that all my modulators are undefined so no automation is possible._
What do you mean by undefined? Do you mean they are not available to control in the DAW?
From what I can see you enabled "Export Parameter to VST host" for each modulator which is good BUT their VST Index are all around "11200" which is a very high ID. I never dealt with more than 1 to 64 so 112XX is really high.
Maybe you should start from "1"  to see if the app pass them properly to the host. Not sure it will work, but it could help.

**Reply by puffer3** [reply]
Thanks @damiensellier 

the whole menu bar is gone so I cant assign midi in or out when an instance is exported.

I will change the VST index id's.  When I copy and paste components in ctrlr it assigns insane large index numbers. 

In ableton with ctrlr plugin the parameters for automation are just 1-64 undefined

![Screen Shot 2022-04-27 at 5 06 33 AM](https://user-images.githubusercontent.com/6006231/165514688-e9990147-b103-4d8f-9900-137e10015a6c.png)

**Comment by damiensellier** [comment]
@puffer3 I never tried my exported instances in Live as AU so I don't know about it.
I just tried loading your panel in Reaper with Ctrlr 5.3.198 and exported it as a restricted AU instance.
And from What I see the menu is shown. I expired my licence of Studio One and don't have Logic so I can't check with those.
Here is what I have:
![OBXCtrlr](https://user-images.githubusercontent.com/55164385/165635723-690421cd-3e75-4a7b-994d-e47973461dce.png)
.

Can you please check if this file is working properly? If it is, you'd better export your AU from another DAW
[OB-SX.component.zip](https://github.com/RomanKubiak/ctrlr/files/8578020/OB-SX.component.zip)

**Reply by puffer3** [reply]
wow this works.  It must be an ableton limitation! I will have to download another DAW!  Thanks damien!

**Reply by puffer3** [reply]
Unfortunately, though your component did work, exporting from Reaper gave me the same menuless output.  Is this a BigSur thing (11.5.2) it seems that nothing works to get around this.

**Reply by damiensellier** [reply]
Hi Henry, that is really strange. 
My Mac Mini is running Mojave 10.14.6 and I exported the AU restricted instance attached above with Ctrlr 5.6.23 in the latest Reaper version v6.56 from April 25 2022
Can you please post your export here so that I can check it in my system to see if it's coming from the AU itself or at a higher level (system...) ?

PS: Let me know if you want me to remove the former link to the export of your panel.

**Reply by puffer3** [reply]
Here it is. your au came up ok. so weird

[Uploading OB-SX.component.zip…]()

**Reply by damiensellier** [reply]
@puffer3 sorry but your upload failed.

**Comment by puffer3** [comment]
sheesh sorry i thought it had finished uploading.  Thanks 

[OB-SX.component.zip](https://github.com/RomanKubiak/ctrlr/files/8596185/OB-SX.component.zip)

**Reply by damiensellier** [reply]
Yes no menubar on this one unfortunately but, you have all the parameters available from the DAW which is good news.
So we can say the bug is coming from the AU itself, corrupted during the export phase
![obx1](https://user-images.githubusercontent.com/55164385/166092948-4a507be5-71e9-4144-8ca0-a0c1d22616d3.jpg)
![obx2](https://user-images.githubusercontent.com/55164385/166092950-4e9b0be1-9033-4b06-9ae2-158bdc775a8e.jpg)
.
Worst case scenario, ask me or someone here to export it for you when you're 100% done ;)

**Reply by puffer3** [reply]
I am making lots of panels so I do want to figure out a way to export that works! lol i have no idea what the problem could be.

**Reply by Godlike-Productions** [reply]
The menu is only available on some versions of ctrlr.  The VST and AU standards heavily discourage menu bars, which I think is why they were removed.  I've never found the little button to be useful for anything except standalones, but there is a command that you can use to show the menu in the versions that support it.

To show menubar:

```panel:getPanelEditor():setProperty("uiPanelMenuBarVisible", 1, false)``` 

To hide menubar:

```panel:getPanelEditor():setProperty("uiPanelMenuBarVisible", 1, false)```

You may also need these commands to enable or disable the menubar on export

```
panel:getPanelEditor():setProperty("uiPanelMenuBarHideOnExport", 0, false)

panel:getPanelEditor():setProperty("uiPanelMenuBarHideOnExport", 1, false)
```

**Reply by damiensellier** [reply]
@Godlike-Productions yeah absolutely. The only required options from the menu are the midi I/O settings which can be added somewhere else like in a popup or another layer, tab etc...

**Reply by puffer3** [reply]
Is there a good panel example where the midi devices and channel have been integrated into the panel?

**Reply by damiensellier** [reply]
The following one has the I/O selector in a modal box:
https://github.com/RomanKubiak/ctrlr/discussions/398#discussioncomment-1916282

Not sure about the midi channel menu...

**Reply by Godlike-Productions** [reply]
We've done them in these panels.

https://godlike.com.au/index.php?id=358

Here's the code.

```
--
-- Called when a mouse is down on this component
--

configure_midi_ports = function(--[[ CtrlrComponent --]] comp, --[[ MouseEvent --]] event)
--if glt.freeze==true then return end
	local myChannelIn=tonumber(panel:getProperty("panelMidiInputChannelDevice"))
	local myInputDevice=panel:getProperty("panelMidiInputDevice")
	local myChannelOut=tonumber(panel:getProperty("panelMidiOutputChannelDevice"))
	local myOutputDevice=panel:getProperty("panelMidiOutputDevice")
	local deviceIdi=0
	local deviceIdo=0
		saMidiInputDevices = StringArray()
		for i=0,(utils.getMidiInputDevices():size()) do
			saMidiInputDevices:set(i, utils.getMidiInputDevices():get(i))
			if utils.getMidiInputDevices():get(i) == myInputDevice then
			deviceIdi=i
			end
		end

		saMidiOutputDevices = StringArray()
		for i=0,(utils.getMidiOutputDevices():size()) do
			saMidiOutputDevices:set(i, utils.getMidiOutputDevices():get(i))
			if utils.getMidiOutputDevices():get(i) == myOutputDevice then
			deviceIdo=i
			end
		end
        saMidiOutputChannel = StringArray()
			for i=1,16 do
			saMidiOutputChannel:set(i,tostring(i))
		end

		saMidiInputChannel = StringArray()
			saMidiInputChannel:set(0,"All channels")
			for i=1,16 do
			saMidiInputChannel:set(i,tostring(i))
		end

		modalWindow = AlertWindow("\nCONFIGURE MIDI PORTS","", AlertWindow.NoIcon)
		modalWindow:addTextBlock ("MIDI INPUT")
		modalWindow:addButton("    OK    ", 1, KeyPress(KeyPress.returnKey),KeyPress())
		modalWindow:addButton("Cancel", 0, KeyPress(KeyPress.escapeKey),KeyPress())
		modalWindow:addButton("Advanced", 2, KeyPress(KeyPress.spaceKey),KeyPress())
		modalWindow:addComboBox ("cmbMidiInputDevices", saMidiInputDevices, "IN Device")
		modalWindow:addComboBox ("cmbMidiInputChannel", saMidiInputChannel, "MIDI Input Channel")
		modalWindow:addTextBlock ("MIDI OUTPUT")
		modalWindow:addComboBox ("cmbMidiOutputDevices", saMidiOutputDevices, "OUT Device")
		modalWindow:addComboBox ("cmbMidiOutputChannel", saMidiOutputChannel, "MIDI Output Channel")


		for i=0,(utils.getMidiInputDevices():size()) do

		end

		modalWindow:getComboBoxComponent("cmbMidiInputDevices"):setSelectedItemIndex(deviceIdi,2)
		modalWindow:getComboBoxComponent("cmbMidiInputChannel"):setSelectedItemIndex(myChannelIn,2)
		modalWindow:getComboBoxComponent("cmbMidiOutputDevices"):setSelectedItemIndex(deviceIdo,2)
		modalWindow:getComboBoxComponent("cmbMidiOutputChannel"):setSelectedItemIndex(myChannelOut-1,2)

		ret = modalWindow:runModalLoop()

		if ret == 1 then

			panel:setPropertyString("panelMidiInputDevice",modalWindow:getComboBoxComponent("cmbMidiInputDevices"):getText())
        	panel:setPropertyInt("panelMidiInputChannelDevice", modalWindow:getComboBoxComponent("cmbMidiInputChannel"):getSelectedItemIndex())
			panel:setPropertyString("panelMidiOutputDevice",modalWindow:getComboBoxComponent("cmbMidiOutputDevices"):getText())
        	panel:setPropertyInt("panelMidiOutputChannelDevice", modalWindow:getComboBoxComponent("cmbMidiOutputChannel"):getSelectedItemIndex()+1)

--local o=panel:getPropertyInt("panelMidiOutputChannelDevice")
--panel:getModulatorByName("rx-channel"):setModulatorValue(o-1,false,false,false)
		end

		if ret == 2 then
		panel:getComponent("midiSettingsButton"):click()
--midiSettingsButton:setModulatorValue(1,false,true,true)
		end --if
		

		modalWindow:setVisible (false)

		end -- f
```

midiSettingsButton is a hidden uiButton with componentInternalFunction set to MIDI Device Selection.  This can only be set on a uiButton, not on a graphic button.

Hope this helps.

**Reply by puffer3** [reply]
Great thanks for this @Godlike-Productions 

Did you create an imageButton that triggers the hidden uibutton?  I have been messing with trying to link the modulators but it doesnt seem very straightforward.

I can create a lua script to link them but that seems like overkill.  Any tips?  Also the advanced button crashes my Ctrlr.

**Reply by Godlike-Productions** [reply]
We created an image button that called the  method above.  There is a hidden UIButton for the MIDISettings that is triggered from the modal box, but it sits on the main panel.  We stuck it in an unused corner.

**Comment by puffer3** [comment]
One other thing I have notices exporting is that now that I have started experimenting with automation, the ctrlr variables seem to have a ton of decimal places (much more than my simple 128 value midi parameters need, or even nrpns i would imagine).  There is an LFO in ableton and it was overloading my CPU doing a simple filter sweep with a sine wave.  There is a max4Live midi lfo that works much better, but i still think that the panel should be able to export in a way that it accepts less complex numbers.  Any ideas?
![Screen Shot 2022-04-30 at 12 24 42 AM](https://user-images.githubusercontent.com/6006231/166096337-37f7f7b2-5f63-400b-89c2-9240872b46ac.jpg)

**Comment by puffer3** [comment]
Things are getting closer now that I can set the midi from within the panel.  

Unfortunately I am losing parameter feedback! Now when I load a patch on my synth and it sends the patch state in CC, the panel wont adjust the modulators to the input value.  The same panel in Ctrlr stand alone does this. I can even see the CC's getting to the panel in the MIDI monitor inside the ctrlr AU in Ableton.  Just no assignment.  Anyone else faced this?

**Reply by damiensellier** [reply]
Not sure if this could help but you can route your input or output from or to the host i/o in the midi settings, you should give it a try :
![io midi host pass through](https://user-images.githubusercontent.com/55164385/167234670-801a3cc4-da8d-4182-ac18-091abc7e2087.png)

When I use several apps like midi monitor, cubase, ctrlr etc on the same external device, I use an IAC midi virtual port  on Mac so that everybody's sharing the same i/o.

**Reply by puffer3** [reply]
I did try these routings but not the IAC port I'll give that a shot!

**Comment by puffer3** [comment, marked as answer]
The reason the file menu wouldnt write with the exports seems to be because I was using version 5.6.23.  when I use 5.3.198 i do get the file menu on export, So there is a reason to keep the older versions on my computer (I have three right now including 5.3.163) which all seem to do different things. 

Ok still trying to get parameter feedback to work! I think I might just implement JSON saving and make all the patch information push instead of push/pull with the synth.  It seems easier to just shove the data at the synth and manage all the patches on the computer!  

Is the Input from host to comparator for reassigning the MIDI channels in the panel? I have tried it but I am still not clear on what the comparator is comparing. 


I do want to figure out why ableton wants to push 8 decimal places to the parameters.  This must make the CPU drain much higher.  Is this a ctrlr bug or an ableton bug ya'll think?  I wonder in what direction the solution lies...

Thanks for all the creative thinking and ideas on these problems.  I have had fun playing my prophet 600 and OBsX (which i have never been able to control) with the depth and control that comes from these panels.

In using @Godlike-Productions MIDI button I was getting dialogues popping up all the time, I must have to try and limit the button with Johns hidden button panel blocker code or something like that.

**Comment by damiensellier** [comment]
I am working on the menu bar not showing up on exports for version 5.6.23+ 
5.6.198 shows/hide menu bar properly, not the newer ctrlr version with juce 6.

**Reply by damiensellier** [reply]
Fixed : 
https://github.com/RomanKubiak/ctrlr/blob/8aa00d82127acda42ad9ac9b7b479461e9436aa4/Source/UIComponents/CtrlrApplicationWindow/CtrlrEditor.cpp#L22

addAndMakeVisible (menuBar = new MenuBarComponent (this));

replaced by 2 different lines as mentioned in the menuBar demo in JUCE : 

```
    menuBar = new MenuBarComponent (this);
    addAndMakeVisible(menuBar);
```

And most important is forcing the visibility if uiPanelMenuBarVisible is set to true : 

https://github.com/RomanKubiak/ctrlr/blob/8aa00d82127acda42ad9ac9b7b479461e9436aa4/Source/UIComponents/CtrlrApplicationWindow/CtrlrEditor.cpp#L64
```

			if ((bool)owner.getActivePanel()->getEditor()->getProperty(Ids::uiPanelMenuBarVisible))
				setMenuBarVisible(true); // set visibility
				editorRect.setHeight (editorRect.getHeight() + (int)owner.getProperty(Ids::ctrlrMenuBarHeight));
```

Hope it helps. I'll go further in the implementation of lookAndFeel for the menuBar since it's broken with the newer JUCE version.

![Capture d’écran, le 2024-01-05 à 05 09 08](https://github.com/RomanKubiak/ctrlr/assets/55164385/e45318f8-101a-4f8d-bdfd-d93bf0a38fb8)
